presto-main/src/main/java/com/facebook/presto/operator/scalar/tdigest/MergingDigest.java

```diff
@@ -0,0 +1,663 @@
+                // centroids i and i+1 bracket our current point
+
+                // check for unit weight
+                double leftUnit = 0;
+                if (weight[i] == 1) {
+                    if (index - weightSoFar < 0.5) {
+                        // within the singleton's sphere
+                        return mean[i];
+                    }
+                    else {
+                        leftUnit = 0.5;
+                    }
+                }
+                double rightUnit = 0;
+                if (weight[i + 1] == 1) {
+                    if (weightSoFar + dw - index <= 0.5) {
+                        // no interpolation needed near singleton
+                        return mean[i + 1];
+                    }
+                    rightUnit = 0.5;
+                }
+                double z1 = index - weightSoFar - leftUnit;
+                double z2 = weightSoFar + dw - index - rightUnit;
+                return weightedAverage(mean[i], z2, mean[i + 1], z1);
+            }
+            weightSoFar += dw;
+        }
+
+        checkArgument(weight[n - 1] > 1, "error");
+        checkArgument(index <= totalWeight, "error");
+        checkArgument(index >= totalWeight - weight[n - 1] / 2, "error");
+
+        // weightSoFar = totalWeight - weight[n-1]/2 (very nearly)
+        // so we interpolate out to max value ever seen
+        double z1 = index - totalWeight - weight[n - 1] / 2.0;
+        double z2 = weight[n - 1] / 2 - z1;
+        return weightedAverage(mean[n - 1], z1, max, z2);
+    }
+
+    @Override
+    public int centroidCount()
+    {
+        mergeNewValues();
+        return activeCentroids;
+    }
+
+    @Override
+    public Collection<Centroid> centroids()
+    {
+        compress();
+        return new AbstractCollection<Centroid>()
+        {
+            @Override
+            public Iterator<Centroid> iterator()
+            {
+                return new Iterator<Centroid>()
+                {
+                    int i;
+
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return i < activeCentroids;
+                    }
+
+                    @Override
+                    public Centroid next()
+                    {
+                        Centroid rc = new Centroid(mean[i], (int) weight[i]);
+                        i++;
+                        return rc;
+                    }
+
+                    @Override
+                    public void remove()
+                    {
+                        throw new UnsupportedOperationException("Default operation");
+                    }
+                };
+            }
+
+            @Override
+            public int size()
+            {
+                return activeCentroids;
+            }
+        };
+    }
+
+    @Override
+    public double getCompressionFactor()
+    {
+        return publicCompression;
+    }
+
+    @Override
+    public int byteSize()
```

Comment:
I think we'll also need to add a method which also estimates the in-memory representation of T-Digest.  See NumericHistogram#estimatedInMemorySize for an example of this